Add TCP TLS guide (#27318)
hashicorp/vault: website/content/docs/configuration/listener/tcp-tls.mdx

ciphersuite selection, which Vault inherits through use of this package.

## Requirements

- A valid TLS certificate file, secret key, and (if required) CA file.

## Default Vault TLS configuration

A default Vault [TCP
listener](/vault/docs/configuration/listener/tcp)
configuration accepts TLS 1.2 or 1.3 connections. Attempts to connect using
TLS 1.0 and 1.1 are dropped.

The ciphersuites used for **TLS 1.2** depends on whether Vault has been
configured with a RSA or ECDSA certificate. The TLS configuration may include
any ciphers supported by Go’s TLS package and Vault’s [ciphersuite configuration
parser](https://github.com/hashicorp/go-secure-stdlib/blob/main/tlsutil/tlsutil.go#L31-L57).

<Highlight title="Go TLS configuration">

While the Go team believes this set of ciphers are appropriate for modern secure
usage, and HashiCorp agrees, some vulnerability scanners may flag issues with
this configuration.

In particular, Sweet32 (CVE-2016-2183) is an attack against 64-bit block size
ciphers including 3DES that may allow an attacker to break the encryption of
long lived connections.

The Go team did not believe this attack was sufficient to remove existing client

> compatibility by deprecating 3DES support (per the researcher’s vulnerability
> disclosure it took a single HTTPS session with 785 GB of traffic to break the

Reviewer: ```suggestion
compatibility by deprecating 3DES support (per the researcher’s [vulnerability
disclosure](https://sweet32.info/) it took a single HTTPS session with 785 GB of
traffic to break the
```

> encryption). Go did deprioritize 3DES in favor of AES-based ciphers.

Reviewer: I think this link is worth bringing through from the draft, as it has additional Go-specific context that may be useful to readers.

Accepted  TLSv1.2  256 bits  AES256-GCM-SHA384
Accepted  TLSv1.2  128 bits  AES128-SHA
Accepted  TLSv1.2  256 bits  AES256-SHA
Accepted  TLSv1.2  112 bits  TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA
Accepted  TLSv1.2  112 bits  TLS_RSA_WITH_3DES_EDE_CBC_SHA

  Server Key Exchange Group(s):
TLSv1.3  128 bits  secp256r1 (NIST P-256)
TLSv1.3  192 bits  secp384r1 (NIST P-384)
TLSv1.3  260 bits  secp521r1 (NIST P-521)
TLSv1.3  128 bits  x25519
TLSv1.2  128 bits  secp256r1 (NIST P-256)
TLSv1.2  192 bits  secp384r1 (NIST P-384)
TLSv1.2  260 bits  secp521r1 (NIST P-521)
TLSv1.2  128 bits  x25519

  SSL Certificate:
Signature Algorithm: sha256WithRSAEncryption
RSA Key Strength:    4096
```

</CodeBlockConfig>

</Tab>
</Tabs>

## Customize Vault TLS configuration

If this default TLS protocol and ciphersuite support is not suitable, Vault’s
TLS configuration can be customized using various parameters within the
[listener](/vault/docs/configuration/listener/tcp) configuration stanza.

### Configure TLS 1.2

If you require TLS 1.2, but with a non-default set of ciphersuites, you can
explicitly define the ciphersuites.

Add the `tls_min_version`, `tls_max_version`, and `tls_ciper_suites` parameters

> to the `listner` stanza.

Reviewer: ```suggestion
to the `listener` stanza.
```

**Example configuration:**

<CodeBlockConfig hideClipboard highlight="5-7">

```plaintext
listener "tcp" {
  address = "127.0.0.1:8200"
  tls_cert_file = "cert.pem"
  tls_key_file  = "key.pem"
  tls_min_version = "tls12"
  tls_max_version = "tls12"

>   tls_cipher_suites = "TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256"}

Reviewer: ```suggestion
  tls_cipher_suites = "TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256,TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256"
}
```

Reviewer: I think it's useful to have the example include multiple ciphersuites as in my draft, rather than limiting to just one.

Author: Thanks - I agree. Was troubleshooting some rendering weirdness and forgot to add it back in. Thanks for the suggestions.

```

</CodeBlockConfig>

Both `tls_min_version` and `tls_max_version` are set to `tls12` to disable usage
of TLS 1.3, as suggested in the [tcp listener
documentation](/vault/docs/configuration/listener/tcp#tls_cipher_suites).

The ciphersuites used in `tls_cipher_suites` are taken from the list supported by
Vault’s ciphersuite configuration parser. The ordering of `tls_cipher_suites` does
not matter, as Go will determine the ordering.

This configuration excludes any 3DES ciphers to avoid potential exposure to the
Sweet32 attack (CVE-2016-2183) described earlier. You can customize this example
to meet environment-specific security requirements.

### Configure TLS 1.3

If a reasonably modern set of clients are connecting to a Vault instance, you
can be configure the tcp listener stanza to accept just TLS 1.3.

Set the `tls_min_version` parameter to `tls13`.

**Example configuration:**

<CodeBlockConfig hideClipboard highlight="5">

```plaintext
listener "tcp" {
  address = "127.0.0.1:8200"
  tls_cert_file = "cert.pem"
  tls_key_file  = "key.pem"
  tls_min_version = "tls13"
}
```

</CodeBlockConfig>

Ciphersuite configuration is not supported in this mode. A select set of ciphers

> has been selected by the Go team and aligned with the [modern

Reviewer: ```suggestion
has been selected by the Go team, which aligns with the [modern
```
